### PR TITLE
[bug#11687] Don't rebuild scala.Seq to drop elems in unapplySeq

### LINKED ISSUE #11687: `LazyList.from(1) match { case LazyList(x, xs @ _*) => xs }` never terminates
Same for `Stream.from(1).view.toSeq` which did terminate in 2.12. The behavior on unapply arises because we go through UnapplySeqWrapper#drop: https://github.com/scala/scala/blob/f23f96c36a3cc0a44a60d28bac0200873686c410/src/library/scala/collection/Factory.scala#L315

/cc @NthPortal, the definitive expert on LazyList :).

### PR DESCRIPTION
**Pull request by NthPortal**
Fixes scala/bug#11687

### CODE REVIEW

File: test/files/run/matchonstream.scala [hunk 1]
```diff
@@ -1,3 +1,3 @@
 object Test extends App{
-  LazyList.from(1) match { case LazyList(1, 2, x @_*) => println("It worked!") }
+  LazyList.from(1) match { case LazyList(1, 2, x @_*) => println(s"It worked! (class: ${x.getClass.getSimpleName})") }
```
Comment (NthPortal) [inline test/files/run/matchonstream.scala:2]: this change to prevent DCE

File: src/library/scala/collection/Factory.scala [hunk 2]
```diff
@@ -312,7 +312,10 @@ object SeqFactory {
     def get: UnapplySeqWrapper[A] = this
     def lengthCompare(len: Int): Int = c.lengthCompare(len)
     def apply(i: Int): A = c(i)
-    def drop(n: Int): scala.Seq[A] = c.view.drop(n).toSeq
+    def drop(n: Int): scala.Seq[A] = c match {
+      case seq: scala.Seq[A] => seq.drop(n)
+      case _                 => c.view.drop(n).toSeq
```
Comment (smarter) [inline src/library/scala/collection/Factory.scala:317]: In which situation is it actually a good idea to go through a view here ?

Comment (NthPortal) [reply src/library/scala/collection/Factory.scala:317]: going through a view avoids building an unnecessary intermediate collection (with the same type as `c`)

Comment (smarter) [reply src/library/scala/collection/Factory.scala:317]: Right, but that's only true for types which aren't Seq but use UnapplySeqWrapper. Which types are those exactly? Do we have test cases that exercise this code path ?

Comment (NthPortal) [reply src/library/scala/collection/Factory.scala:317]: we don't. and they will still be forced.

I think saying that it ought to be `LazyList.from(c.view.drop(n))` (or `LazyList.from(c).drop(n)`) to preserve laziness is valid

Comment (NthPortal) [reply src/library/scala/collection/Factory.scala:317]: I can hear @dwijnand's voice in my head asking if this is really all necessary and the best design

Comment (julienrf) [reply src/library/scala/collection/Factory.scala:317]: > Right, but that's only true for types which aren't Seq but use UnapplySeqWrapper. Which types are those exactly? 

All the `mutable.Seq` leaves.

**Review by julienrf — APPROVED** [review]

File: test/files/run/matchonstream.scala (hunk 1 above)
Comment (julienrf) [reply test/files/run/matchonstream.scala:2]: Does that mean that the `match` expression was not even evaluated because of DCE?

Comment (NthPortal) [reply test/files/run/matchonstream.scala:2]: I think, based on @smarter's comment, just the `x @_*` (which is `UnapplySeqWrapper#drop`) wasn't evaluated?

Comment (szeiger) [reply test/files/run/matchonstream.scala:2]: patmat never generates the code in the first place if the symbol is unused.

**Review by szeiger — APPROVED** [review]
This is a lot simpler than what I had in mind.

**Comment by julienrf** [comment]
Thanks @NthPortal :+1:

**Pull request merged**